### PR TITLE
Kakfa Alert Scheme for Notifications

### PR DESCRIPTION
**Pull request by apoorvmittal10**

### CODE REVIEW

File: thirdeye/pom.xml [hunk 1]
```diff
@@ -79,7 +79,8 @@
     <jsonpath.version>2.4.0</jsonpath.version>
     <!-- force bump up for LI usage -->
     <glassfish.hk2.utils.version>2.5.0-b34</glassfish.hk2.utils.version>
-    <kafka.clients.version>0.10.2.1</kafka.clients.version>
```
Comment (devinay) [inline thirdeye/pom.xml:82]: Do all the tests work ? 
Will the calcite and the avatica dependencies be on another PR ?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java [hunk 2]
```diff
@@ -77,6 +79,10 @@ public static BaseNotificationContent buildNotificationContent(Properties alertS
         content =  new EntityGroupKeyContent();
         break;
 
+      case KAFKA:
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:82]: Is there any specific reason to override in the base class  instead of restricting all changes to DetectionKafkaAlerter ?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java [hunk 3]
```diff
@@ -77,6 +79,10 @@ public static BaseNotificationContent buildNotificationContent(Properties alertS
...
+        content =  new KafkaNotificationContent();
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:83]: Hmm looks like creating a Kafka Notification without overriding the base is too much unneeded work, thats the reason for overriding the base class.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java [hunk 4]
```diff
@@ -77,6 +79,10 @@ public static BaseNotificationContent buildNotificationContent(Properties alertS
...
+        break;
+
       default:
         throw new IllegalArgumentException(String.format("Unknown email template '%s'", template));
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:87]: Minor nit : it becomes "Unknown template" as its not email.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionKafkaAlerter.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ * alertSchemes:
+ * - type: KAFKA
+ *   params:
+ *     name: "kafka-subscription-integration"           # Should be same as 'name' defined in al
+ */
+@AlertScheme(type = "KAFKA")
+public class DetectionKafkaAlerter extends DetectionAlertScheme {
+  private static final Logger LOG = LoggerFactory.getLogger(DetectionKafkaAlerter.class);
+  public static final String PROP_KAFKA_SCHEME = "kafkaScheme";
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionKafkaAlerter.java:78]: Would "kafkaAlertScheme" be clearer, since Kafka is used elsewhere in the repo for different functions as well.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/commons/KafkaEntity.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  public void setMetadata(final Map<String, Object> metadata) {
+    this.metadata = metadata;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  @Override
+  public String toString() {
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/commons/KafkaEntity.java:138]: Could we use JSON and some standard ObjectMapper ?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/content/templates/KafkaNotificationContent.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.pinot.thirdeye.notification.content.BaseNotificationContent;
+import org.apache.pinot.thirdeye.rootcause.impl.MetricEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This kafka formatter for notifications content.
+ */
+public class KafkaNotificationContent extends BaseNotificationContent {
```
Comment (devinay) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/content/templates/KafkaNotificationContent.java:46]: Could we model this as a first class JSON Object ?

File: thirdeye/pom.xml (hunk 1 above)
Comment (apoorvmittal10) [reply thirdeye/pom.xml:82]: Test cases fail generally for me: https://apache-thirdeye.slack.com/archives/C014CS60A22/p1594655051055200

Druid calcite integration will change current code hence current test cases success more important in that PR.

Yes, druid calcite PR will be different.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java (hunk 2 above)
Comment (apoorvmittal10) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:82]: I am not clear specifically what you meant. Please if you can help me here?

Though `BaseNotificationContent` class is abstract which is implemented by 3 other classes `EntityGroupKeyContent`, `HierarchicalAnomaliesContent` and `MetricAnomaliesContent`. 3 of the classes try to fill a lot of data specific to anomalies, trying to take error screenshot, etc. I just needed a basic template for kafka with minimal capabilities as of now. May be later if we need to have more details then discarding `KafkaNotificationContent` ll make sense.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java (hunk 3 above)
Comment (apoorvmittal10) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:83]: `public class KafkaNotificationContent extends BaseNotificationContent {` Again seems I am missing something to understand here, please if you can help.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionKafkaAlerter.java (hunk 5 above)
Comment (apoorvmittal10) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionKafkaAlerter.java:78]: Actually this is the variable name `thirdeye` injects in db row `implicitly` as per alert scheme, so can't help :) 

https://github.com/confluentinc/incubator-pinot/blob/2ec7dee1597021742f68f0ae8b279f7560e55894/thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/translator/SubscriptionConfigTranslator.java#L158

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/commons/KafkaEntity.java (hunk 6 above)
Comment (apoorvmittal10) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/commons/KafkaEntity.java:138]: Yeah, completely make sense. `ObjectMapper` can be used, I just followed the way they did in other classes. I should change it to use `ObjectMapper`.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/content/templates/KafkaNotificationContent.java (hunk 7 above)
Comment (apoorvmittal10) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/notification/content/templates/KafkaNotificationContent.java:46]: I think you mean `KafkaEntity` class which is being written to `kafka` shall be first class json, is it?

**Comment by apoorvmittal10** [comment]
@devinay I have tried to add comments and details for review. Please if you can help.

> We could then optimize away KafkaContentFormatter and always use JSON whenever we hit the wire (kafka)

I think you meant `KafkaEntity` (pojo class) shall be first class JSON as that is being written to `kafka`. The idea is to use `JsonSerializer`, `ByteSerializer` or etc as per convenience to write content to kafka in desired format. The code change which shall happen has to be regarding this class shall be in some schema format like `protobuf` or `avro` which will make it easy to share to client in future.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java (hunk 2 above)
Comment (devinay) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/scheme/DetectionAlertScheme.java:82]: > I am not clear specifically what you meant. Please if you can help me here?

You have addressed my concern, I was wondering why we did not create a JSONKafkaContent class and not change the base class, then we could have used a standard ObjectMapper to convert from JSON to the POJO.

If we are going to go with a serialization framework like Avro or Protobuf, this is okay too.
